### PR TITLE
[release/5.0] When marshalling a layout class, fall-back to dynamically marshalling the type if it doesn't match the static type in the signature.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Port of #50137, #50655, and #50735 to release/5.0

When we switched the struct marshalling system to be IL based, we accidentally stopped checking if the passed in object matched the type in the signature in the "layout class" case (where we have a sequential or explicit layout `class` type). This causes failures to marshal the derived types correctly since only the base type ends up being marshalled.

## Customer Impact

Without this change, developers need to manually marshal their types. An internal team (Halo) has worked around this by manually marshalling their types, but would prefer to remove this workaround.

## Regression?

This was a regression from .NET Core 3.1.

See #49857

## Testing

Tests for both the fix and backwards compatibility are included with this change.

## Risk

Medium. This change affects marshalling of layout classes.

### CODE REVIEW

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Review by jeffschwMSFT — APPROVED** [review]
Approved.  I will take for consideration in 5.0.x.

**Comment by jkotas** [comment]
No merge until https://github.com/dotnet/runtime/issues/50410 is fixed

**Review by jeffschwMSFT — APPROVED** [review]
Approved.  Please complete cr.

**Comment by jeffschwMSFT** [comment]
@leecow we had approved this issue.  Then in the time since a customer came back with an issue in master that applies to this change.  We have updated the PR (the original intent of the change remains the same).  I suggest that we consider this issue approved based on our original discussion.  Though happy to bring it back if further discussion is warranted.

**Comment by jeffschwMSFT** [comment]
@jkoritzinsky is this PR now in a mergeable state?

**Comment by jkoritzinsky** [comment]
Yes, this PR is in a mergable state. Just waiting on an updated code review @AaronRobinsonMSFT

**Comment by jeffschwMSFT** [comment]
@Anipik once we have @AaronRobinsonMSFT cr, this pr will be ready for merge.  We found an issue and have sense included an updated commit to address it.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by jeffschwMSFT** [comment]
@Anipik this pr is ready to merge.  Thanks

**Comment by Anipik** [comment]
> @Anipik this pr is ready to merge. Thanks

i will merge the pr after the ci is complete.

**Pull request merged**
